syntax = "proto3";

package raft;

// RequestVote RPC - used during leader election
message RequestVoteRequest {
    int32 term = 1; // candidate's term
    string candidate_id = 2; // candidate requesting vote
    int32 last_log_index = 3; // index of candidate's last log entry
    int32 last_log_term = 4; // term of candidate's last log entry
}

message RequestVoteResponse {
    int32 term = 1; // current term, for candidate to update itself
    bool vote_granted = 2; // true if vote granted
}

// AppendEntries RPC - used for log replication and heartbeat
message LogEntry {
    int32 term = 1; // term when entry was received by leader
    string command = 2; // command for state machine (key-value operation)
    int32 index = 3; // log index
}

message AppendEntriesRequest {
    int32 term = 1; // leader's term
    string leader_id = 2; // so follower can redirect clients
    int32 prev_log_index = 3; // index of log entry immediately preceding new ones
    int32 prev_log_term = 4; // term of prev_log_index entry
    repeated LogEntry entries = 5; // log entries to store (empty for heartbeat)
    int32 leader_commit = 6; // leader's commit_index
}

message AppendEntriesResponse {
    int32 term = 1; // current term, for leader to update itself
    bool success = 2; // true if follower contained entry matching prev_log_index and prev_log_term
    int32 conflict_index = 3; // for faster log backtracking on conflict
    int32 conflict_term = 4; // for faster log backtracking on conflict
}

// Client request to add a command to the log
message ClientRequest {
    string command = 1; // command to execute (e.g., "SET key value")
}

message ClientResponse {
    bool success = 1; // true if command was committed
    string message = 2; // status message or error
    string leader_id = 3; // current leader's ID (for redirection)
}

// Special RPC for network partition testing
message IsolateRequest {
    repeated string isolated_nodes = 1; // list of node IDs to isolate from
}

message IsolateResponse {
    bool success = 1;
    string message = 2;
}

// Service definition
service RaftService {
    // Core RAFT RPCs
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    
    // Client interaction
    rpc SubmitCommand(ClientRequest) returns (ClientResponse);
    
    // Testing utilities
    rpc Isolate(IsolateRequest) returns (IsolateResponse);
}
